src/logging.c

```diff
@@ -628,12 +667,131 @@ rcutils_ret_t rcutils_logging_format_message(
   return status;
 }
 
+#ifdef WIN32
+# define COLOR_NORMAL 7
+# define COLOR_RED 4
+# define COLOR_GREEN 2
+# define COLOR_YELLOW 6
+#else
+# define COLOR_NORMAL "\033[0m"
+# define COLOR_RED "\033[31m"
+# define COLOR_GREEN "\033[32m"
+# define COLOR_YELLOW "\033[33m"
+#endif
+
+#define SET_COLOR_WITH_SEVERITY(status, severity, color) \
+  { \
+    switch (severity) { \
+      case RCUTILS_LOG_SEVERITY_DEBUG: \
+        color = COLOR_GREEN; \
+        break; \
+      case RCUTILS_LOG_SEVERITY_INFO: \
+        color = COLOR_NORMAL; \
+        break; \
+      case RCUTILS_LOG_SEVERITY_WARN: \
+        color = COLOR_YELLOW; \
+        break; \
+      case RCUTILS_LOG_SEVERITY_ERROR: \
+      case RCUTILS_LOG_SEVERITY_FATAL: \
+        color = COLOR_RED; \
+        break; \
+      default: \
+        fprintf(stderr, "unknown severity level: %d\n", severity); \
+        status = RCUTILS_RET_INVALID_ARGUMENT; \
+    } \
+  }
+#ifdef WIN32
+# define IS_OUTPUT_COLORIZED(is_colorized, stream) \
+  { \
+    if (g_colorized_output == RCUTILS_COLORIZED_OUTPUT_FORCE_ENABLE) { \
+      is_colorized = true; \
+    } else if (g_colorized_output == RCUTILS_COLORIZED_OUTPUT_FORCE_DISABLE) { \
+      is_colorized = false; \
+    } else { \
+      is_colorized = _isatty(_fileno(stream)) != 0; \
```

Comment:
I don't know if `_isatty` is the right thing to use here, because it returns non-zero when the filedescriptor is a character device (and not a terminal like in linux/mac).
I tried some examples without seeing a `SetConsoleTextAttribute` error, so I'm happy with it.
- In the following example, I see a `SetConsoleTextAttribute` error if I force colorizing the output with the environment variable. It doesn't appear with the auto behavior.
```bat
ros2 run PACKAGE NODE_THAT_LOGS __log_level:=debug > log.txt
```
- I haven't seen errors when redirecting `stdout` to other character devices. e.g.:
```bat
ros2 run PACKAGE NODE_THAT_LOGS __log_level:=debug > null
```

Option 2:
- Use `isatty`, and add a pragma for avoiding the deprecated warning.

@dirk-thomas What do you think?

Reply:
I don't know enough about `_isatty` but sounds good to me if it works for you..